indexer: harden txn response type for type safety

gegaowp opened the pull request:
## Description 

- change txn response type on indexer end to SuiTransactionFullResponse, which has no unexpected None values; If unexpected None happens, indexer will retry and also log errors to k8 pod.
- other minor changes like always use ms rather than naiveDateTime

## Test Plan 

- CI
- local test run with local validator

wlmyng commented on crates/sui-indexer/tests/integration_tests.rs:
```diff
@@ -2,7 +2,7 @@
 
 // integration test with standalone postgresql database
-#[cfg(feature = "pg_integration")]
+// #[cfg(feature = "pg_integration")]
```
nit: should uncomment this

wlmyng commented on crates/sui-sdk/examples/transfer_coins.rs:
```diff
@@ -4,14 +4,10 @@
-    },
-    SuiClientBuilder,
-};
+use sui_sdk::SuiClientBuilder;
```
question: what's the diff between sui_sdk::types and sui_types?

gegaowp replied:
sui_types are basic sui protocol types, sui-json-rpc-types are user facing ones, we did not have sdk types until this one newly added one, changes are needed here b/c of types overriding.

wlmyng reviewed (approved):
:mgtlskje:

666lcz commented on crates/sui-sdk/src/types.rs:
```diff
@@ -0,0 +1,132 @@
+    pub object_changes: Vec<ObjectChange>,
+    pub balance_changes: Vec<BalanceChange>,
+    pub timestamp_ms: u64,
+    pub confirmed_local_execution: bool,
```
This is expected to be None for `get_transactions`

gegaowp replied:
can this still be None while running multi_get_transactions_with_options with full_content() after checkpoint is processed?

666lcz reviewed (changes requested):
I disagree with the approach here, let's sync offline

gegaowp commented:
synced offline, the concern about change in sui-sdk is valid, I moved it to indexer cc @666lcz

666lcz commented on crates/sui-indexer/src/handlers/checkpoint_handler.rs:
```diff
@@ -96,6 +97,10 @@ where
             // a checkpoint not generated yet.
             while checkpoint.is_err() {
+                error!(
+                    "Failed to download checkpoint data with error {:?}, retrying...",
```
nit: we might still want to log `next_cursor_sequence_number` since `checkpoint` might not contain the id when there's an error

666lcz commented on crates/sui-indexer/src/handlers/checkpoint_handler.rs:
```diff
@@ -149,37 +148,30 @@ where
-            .await?;
+            .await
+            .map_err(|e| {
+                IndexerError::FullNodeReadingError(format!("Failed to get checkpoint: {:?}", e))
```
nit: log seq

gegaowp commented on crates/sui-indexer/src/types.rs:
```diff
@@ -0,0 +1,132 @@
+        } = response;
+
+        let transaction = transaction.ok_or_else(|| {
+            anyhow::anyhow!(
```
another argument around this is, when we see unexpected None values, this way we know which field is missing, but yeah I did not marry these 50 LOCs, and happy to be convinced. cc @patrickkuo

666lcz commented on crates/sui-indexer/src/handlers/checkpoint_handler.rs:
```diff
@@ -228,6 +220,12 @@ where
+            })
+            .map_err(|e| {
+                IndexerError::SerdeError(format!(
+                    "Failed to generate changed objects with err {:?}",
```
log `tx.digetst`

gegaowp replied:
this is about all objects of a checkpoint, logging checkpoint seq instead.

666lcz commented on crates/sui-indexer/src/handlers/checkpoint_handler.rs:
```diff
@@ -382,17 +363,16 @@ where
-                    .cloned()
-                    .chain(effects.mutated().iter().cloned())
-                    .chain(effects.unwrapped().iter().cloned())
+                let created = tx.effects.created().iter().cloned();
+                let mutated = tx.effects.mutated().iter().cloned();
+                let unwrapped = tx.effects.unwrapped().iter().cloned();
```
nit: I think you can avoid clone here

666lcz commented on crates/sui-indexer/src/lib.rs:
```diff
@@ -187,3 +192,29 @@ pub async fn build_json_rpc_server<S: IndexerStore + Sync + Send + 'static + Clo
+        .await
+        .map_err(|e| {
+            IndexerError::FullNodeReadingError(format!(
+                "Failed to get transactions with error: {:?}",
```
log digest

666lcz commented on crates/sui-indexer/src/lib.rs:
```diff
@@ -187,3 +192,29 @@ pub async fn build_json_rpc_server<S: IndexerStore + Sync + Send + 'static + Clo
+        .collect::<Result<Vec<_>, _>>()
+        .map_err(|e| {
+            IndexerError::FullNodeReadingError(format!(
+                "Unexpected None value in SuiTransactionFullResponse {:?}",
```
log digests

666lcz reviewed (commented):
Overall LGTM! Just some nits about logging to improve debuggability

666lcz commented on crates/sui-indexer/src/types.rs:
```diff
@@ -0,0 +1,132 @@
+                digest
+            )
+        })?;
+        let confirmed_local_execution = confirmed_local_execution.ok_or_else(|| {
+            anyhow::anyhow!(
+                "ConfirmedLocalExecution is None in SuiTransactionFullResponse of digest {:?}.",
+                digest
+            )
+        })?;
```
this is expected to be None though, I would keep this as `Option`

666lcz commented on crates/sui-indexer/src/types.rs:
```diff
@@ -0,0 +1,132 @@
+            timestamp_ms,
+            confirmed_local_execution,
+            checkpoint,
+            errors: _,
```
you probably want to return error if `errors` is not empty

gegaowp commented:
made several other changes:
- turns out multi get do not include obj & balance changes, removed them
- move checkpoint while to download such that waiting for new checkpoint will not generate noisy errors
- addressed Chris's comments about errors and local execution
